src/components/hv-date-field/index.js

```diff
@@ -0,0 +1,321 @@
+import type { Node as ReactNode } from 'react';
+import type { StyleSheet as StyleSheetType } from 'react-native/Libraries/StyleSheet/StyleSheetTypes';
+import styles from './styles';
+
+/**
+ * A date field renders a form field with ISO date fields (YYYY-MM-DD).
+ * Focusing the field brings up a system-appropriate UI for date selection:
+ * - On iOS, pressing the field brings up a custom bottom sheet with a picker and action buttons.
+ * - On Android, pressing the field brings up the system date picker modal.
+ */
+export default class HvDateField extends PureComponent<Props, State> {
+  static namespaceURI = Namespaces.HYPERVIEW;
+  static localName = LOCAL_NAME.DATE_FIELD;
+  static contextType = FormatDateContext;
+  props: Props;
+  state: State;
+
+  constructor(props: Props) {
+    super(props);
+    const element: Element = props.element;
+    const stringValue: ?DOMString = element.getAttribute('value');
+    const value: ?Date = this.createDateFromString(stringValue);
+    const pickerValue: Date = value || new Date();
+    this.state = {
+      // Date that's selected in the field. Can be null.
+      value,
+      // Date shown when the picker opens, must be set to a default to display.
+      pickerValue,
+      focused: false,
+      fieldPressed: false,
+      donePressed: false,
+      cancelPressed: false,
+    };
+  }
+
+  /**
+   * Given a ISO date string (YYYY-MM-DD), returns a Date object. If the string
+   * cannot be parsed or is falsey, returns null.
+   */
+  createDateFromString = (value: ?string): ?Date => {
+    if (!value) {
+      return null;
+    }
+    const [year, month, day] = value.split('-').map(p => parseInt(p, 10));
+    return new Date(year, month - 1, day);
+  };
+
+  /**
+   * Given a Date object, returns an ISO date string (YYYY-MM-DD). If the Date
+   * object is null, returns an empty string.
+   */
+  createStringFromDate = (date: ?Date): string => {
+    if (!date) {
+      return '';
+    }
+    const year = date.getUTCFullYear();
+    const month = date.getUTCMonth() + 1;
+    const day = date.getUTCDate();
+    return `${year}-${month}-${day}`;
+  };
+
+  toggleFieldPress = () => {
+    this.setState({ fieldPressed: !this.state.fieldPressed });
+  };
+
+  toggleCancelPress = () => {
+    this.setState({ cancelPressed: !this.state.cancelPressed });
+  };
+
+  toggleSavePress = () => {
+    this.setState({ donePressed: !this.state.donePressed });
+  };
+
+  /**
+   * Shows the picker, defaulting to the field's value.
+   */
+  onFieldPress = () => {
+    this.setState({
+      focused: true,
+    });
+  };
+
+  /**
+   * Hides the picker without applying the chosen value.
+   */
+  onModalCancel = () => {
+    this.setState({
+      focused: false,
+    });
+  };
+
+  /**
+   * Hides the picker and applies the chosen value to the field.
+   */
+  onModalDone = () => {
+    const element: Element = this.props.element;
+    this.setState({
+      focused: false,
+      value: this.state.pickerValue,
```

Comment:
Minor: since this is a derived state as well, it should use a callback pattern to update state as well.